### PR TITLE
thenReply Effect, #25482

### PR DESCRIPTION
**Pull request by patriknw**
Refs #25482

### CODE REVIEW

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorSpec.scala [hunk 1]
```diff
@@ -148,6 +150,10 @@ object PersistentBehaviorSpec {
         case IncrementWithPersistAll(n) ⇒
           Effect.persist((0 until n).map(_ ⇒ Incremented(1)))
 
+        case cmd: IncrementWithConfirmation ⇒
+          Effect.persist(Incremented(1))
+            .thenReply(cmd)(newState ⇒ Done)
```
Comment (chbatey) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorSpec.scala]: I'm a :-1: to creating an alias for sending a message

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala [hunk 2]
```diff
@@ -48,7 +53,13 @@ object ClusterShardingPersistenceSpec {
       entityId,
       emptyState = "",
       commandHandler = (state, cmd) ⇒ cmd match {
-        case Add(s) ⇒ Effect.persist(s)
+        case Add(s) ⇒
+          Effect.persist(s)
+
+        case cmd @ AddWithConfirmation(s) ⇒
+          Effect.persist(s)
+            .thenReply(cmd)(newState ⇒ Done)
```
Comment (johanandren) [inline akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: Compared to the current zero-api 
```
.thenRun(newState => 
  cmd.replyTo ! Done
)
```

I don't think this provides any value at all. You could just as easily forget to do the `thenReply`

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorSpec.scala (hunk 1 above)
Comment (hepin1989) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorSpec.scala]: Yes,why not just a method named `ackDone`

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala (hunk 2 above)
Comment (viktorklang) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: Well, it does provide the value of allowing for at least an improved discoverability on *where* to issue replies.

Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: That is true. On the other hand it requires some ceremony when defining the commands; have to implement a trait compared to defining the `replyTo` in the messages in whatever way you like. It also adds another way to send messages. I'm on the fence but would be ok with having this if others think it's needed.

Comment (TimMoore) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: Unless there can be some kind of type checking that a reply is issued at all, I agree with @johanandren and @chbatey that this doesn't seem valuable.

I'm still clinging to a hope that there's some kind of (opt-in) pattern possible that could type check the reply (including the presence of a reply), at least for simple request-response flows that I think are the common case for persistent entities.

Comment (viktorklang) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: But we can enforce the response, by requiring a specific type of effect to be returned: a ResponseEffect.

That is orthogonal though. `thenTell` is another option. I've seen cases where people respond *outside* of the effects, and that's not great.

Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: > requiring a specific type of effect to be returned: a ResponseEffect

Could you show me how to do that? That return type depends on the specific command type.

>  I've seen cases where people respond outside of the effects, and that's not great.

If command is not accepted you want to respond outside an effect, in my opinion. I see no point in doing that with `Effect.none.thenReply(cmd)(_ => Rejected)` when all it is `cmd.replyTo ! Rejected`.

There can be all kinds of variations, perhaps you want to validate the command by interacting with some other actors before replying. Perhaps you also want to sent invalid commands to the overlord for audit.

Comment (viktorklang) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: >Could you show me how to do that? That return type depends on the specific command type.

```scala
sealed trait Cmd { type Response }

final case class AddPorsche(licensePlateNo: String) extends Cmd { type Response = Boolean }

final case class ShowPorsche(licensePlateNo: String) extends Cmd { type Response = String }

trait ResponseEffect

object Effects {
  def respondTo[T <: Cmd](command: T)(response: command.Response): ResponseEffect = ???
}

trait PersistentEntity {
  def onCommand(cmd: Cmd) : ResponseEffect
}

class PorscheCollector extends PersistentEntity {
  override def onCommand(cmd: Cmd): ResponseEffect =
    cmd match {
      case a @ AddPorsche(plate) => Effects.respondTo(a)(true)
      case s @ ShowPorsche(plate) => Effects.respondTo(s)("A really nice Porsche")
    }
}
```

Comment (viktorklang) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: @patriknw `Effects.noResponse` and have those Cmd's which do not respond use NotUsed?

Comment (patriknw) [reply akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: That's a good suggestion. Enforcing it on all commands with explicit opt-out is possible.
I have implemented that and pushed to this PR, see PersistentBehaviorReplySpec: https://github.com/akka/akka/pull/25692/files#diff-deb775cb1ed6e5e768717073cc78711eR37

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehavior.scala [hunk 3]
```diff
@@ -45,6 +46,16 @@ object PersistentBehavior {
     eventHandler:   (State, Event) ⇒ State): PersistentBehavior[Command, Event, State] =
     PersistentBehaviorImpl(persistenceId, emptyState, commandHandler, eventHandler)
 
+  /**
+   * Create a `Behavior` for a persistent actor that is enforcing replies to commands.
+   */
+  def withEnforcedReplies[Command <: ExpectingReply[_], Event, State](
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehavior.scala]: Note that the non-enforcing apply remains as an alternative. I don't want to make the enforcement of replies always mandatory because it's not essential for the minimal core API, and it dictates of commands must be defined with the ExpectingReply type.

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+          Effect.persist(Incremented(1))
+            .thenRun((_: State) ⇒ ctx.self ! ReplyNow(cmd.replyTo))
+            .thenNoReply(cmd)
+
+        case cmd: ReplyNow ⇒
+          Effect.replyDone(cmd)
+
+        case query: GetValue ⇒
+          Effect.reply(query)(state)
+
```
Comment (patriknw) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: note that this only compiles if a `ReplyEffect` is returned, which can be created with `Effect.reply`, `Effect.noReply`, `Effect.replyDone`, `.thenReply`, `.thenNoReply`, `.thenReplyDone`.

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  final case class State(value: Int, history: Vector[Int])
+
+  def counter(persistenceId: PersistenceId)(implicit system: ActorSystem[_]): Behavior[Command[_]] =
+    Behaviors.setup(ctx ⇒ counter(ctx, persistenceId))
+
+  def counter(
+    ctx:           ActorContext[Command[_]],
+    persistenceId: PersistenceId): PersistentBehavior[Command[_], Event, State] = {
+    PersistentBehavior.withEnforcedReplies[Command[_], Event, State](
```
Comment (chbatey) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: Nice. I like that you decide once then get the compiler to check you've replied.

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala (hunk 4 above)
Comment (debasishg) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: Looks like a message based implementation of phantom types .. nice!

Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: lol, I have always looked for a use case for those phantoms

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala [hunk 6]
```diff
@@ -58,6 +63,18 @@ object Effect {
...
    */
   def stop[Event, State]: Effect[Event, State] = none.andThenStop()
+
+  // FIXME doc all reply and thenReply stuff
+
+  def reply[ReplyMessage, Event, State](cmd: ExpectingReply[ReplyMessage])(replyWithMessage: ReplyMessage): ReplyEffect[Event, State] =
+    none[Event, State].thenReply[ReplyMessage](cmd)(_ ⇒ replyWithMessage)
+
+  def replyDone[Event, State](cmd: ExpectingReply[Done]): ReplyEffect[Event, State] =
+    reply(cmd)(Done)
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala]: Now the user may be in the other, default mode, and accidentally use these and think there will be an automagic reply. So we have gone from having the risk of missing out on sending a reply, to having the risk of thinking that you replied but did in fact not?

Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala]: nope, there are no automagic replies here, these effects will also work when you use them in a non-enforcing PersistentBehavior

Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala]: Ah, good, it's just that they aren't enforced. Got it 👍

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala (hunk 4 above)
Comment (WadeWaldron) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: How will it work if I want the reply to come from another actor?

Also, `.noReply` and `.thenNoReply` seem like odd options given that we are trying to force a reply.

Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: Exactly, it's conscious opt-out decision. Reason could be things like:
* not all commands require a reply
* a reply is sent later in the process, e.g. after some interaction with another actor

Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: > How will it work if I want the reply to come from another actor?

In the end it's just about sending a message to the `replyTo: ActorRef` in the command, so you can carry on that to another actor or keep it (or the original command) around until it's time to reply.

That is somewhat illustrated in this example with the `IncrementReplyLater` that is sending a `ReplyNow` to self, which is then sending the reply.

**Review by octonato — COMMENTED** [review]
This is really looking great, but I'm biased because it's very close to what I have been experimenting. :-)

I have done almost the same but then using a `HandlerFactory` similar to what I have done for Optional state. 

I hit one roadblock though, that is also present here, but we are not seeing it because that new API wasn't exercised enough. 

How do I fail a command as such that the `Future` returned by `ask` fails as well?

**Comment by renatocaval** [comment]
To give some more context, these are the error types I was defining on my own experiment.

```scala
sealed trait Response

sealed trait Confirmation extends Response
case object Accepted extends Confirmation
case class Rejected(reason: String) extends Confirmation

sealed trait Query extends Response
case class CurrentBalance(amount: Double) extends Query
```

The response types are part of the model protocol and are defined together with the commands and events. 

If I decide to reject a command, I can simply do:

```scala
cmd.replyTo ! Rejected("Invalid command")
```

The problem I see here is that the called will receive a successful `Future[Confirmation]` with a `Rejected`.

When using such a Future in a for-comprehension or flatMap, we will get unexpected results. 

```scala
for {
  res <- accountRef ? Withdraw(100)
  _ <- notifySomeone()
} yield ...
```

It will more intuitive if they get a failed `Future` is an `InvalidCommand` exception. 

I think that the only way to solve it is to have a marker trait in Akka, something like: `trait Error { def reason: String }`. Then the `ask` would need to be a little smarter and fail the future if the response extends it. 

Of course, users are free to use it or not. They can also force the future to a failure by hand, but that's on them to choose. 

WDYT?

**Comment by patriknw** [comment]
> How do I fail a command as such that the Future returned by ask fails as well?

The Future of an ask should not be failed for rejections. That should be part of the message protocol, i.e. that you can get an acknowledgment or a rejection. Perhaps we should define a data type for that, which is more useful than `Done` for these interactions?

The only case when the ask Future may fail is for timeouts.

In untyped one could send back a special `Status.Failure` to `sender()`, but that isn't possible in Typed.

Another aspect is the commands can be sent via `ActorRef.tell` or `ActorContext.ask` and then there is no `Future` so it'a better to have ordinary messages also for rejections.

BTW, exceptions shouldn't be used for rejections.

**Comment by patriknw** [comment]
such `trait Error` could work

wdyt @viktorklang @johanandren @chbatey ?

**Comment by renatocaval** [comment]
@patriknw, I agree. The failure must be part of the protocol. 

The idea of `trait Error` is a compromise to bring back the old behavior in untyped where one could fail the `Future` returned by an `ask`.  

What I'm liking in that, is that we can define a whole hierarchy of types to express responses. Incluvise a way to force a failed `Future`. And the user can choose.

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala (hunk 4 above)
Comment (WadeWaldron) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: When I read this: `withEnforcedReplies` that seems to imply to me that we are enforcing that a reply happen. Looking at it from the outside I would be tempted to assume that if someone has implemented `withEnforcedReplies` then I can always expect that I will receive a reply.

That's obviously not true here. In fact, `withEnforcedReplies` actually doesn't give you any more guarantees about whether a reply is being returned than you had when it wasn't there. The only thing it is doing is making it slightly harder to write bad code. Although, if you are likely to forget to reply, then you are probably also likely to forget to include the `withEnforcedReplies` trait. There is maybe a small amount of additional security once it's there, but there is still a lot of room to mess things up, and the trait adds additional complexity (new ways to reply etc).

So I guess we have to ask what our goal is here. Is our goal to make it harder to introduce a bug in the code? If that's the case, then we see a small amount of success with this. If our goal is to provide a more intent revealing API, or one that is simpler and easier to use, then I think this fails to meet that criteria.

**Comment by viktorklang** [comment]
@renatocaval @patriknw I'm not sure I understand the question, is there a full example?

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala (hunk 4 above)
Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: The primary goal was to help was to detect that reply was forgotten, avoid mistakes.

I didn't put much thinking into the name `withEnforcedReplies`, so open for suggestions.

> small amount of additional security once it's there

In what way is it a small amount? It will not compile if you don't.

Comment (WadeWaldron) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: It's small because there is still ways to mess it up. Assuming I remember to put the trait in, I could still use the wrong reply type (eg. replyDone instead of reply etc.). But the real reason I say it is small is because I am not sure that forgetting the reply is a huge problem. It's something that happens on occasion, but likely gets caught relatively quickly.

This is honestly not a problem I see a lot of people complaining about. I am not sure I have encountered it on an engagement. I have probably done it a few times in my own code, but then realized my mistake very quickly and fixed it.

I will say though that I don't see many support cases. So maybe it's a problem that comes up often on support cases.

Comment (WadeWaldron) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorReplySpec.scala]: Ah yes, I see now that you can't in fact use the wrong reply type. Thanks.

If forgetting the reply is a huge problem in Lagom, it's not one that I have seen. So as I said, it must be something that comes up in support cases.

**Comment by renatocaval** [comment]
@viktorklang, I will put it on a PR. 

But to give a short preview:

### The problem

```scala
sealed trait Confirmation extends Response
case object Accepted extends Confirmation
case class Rejected(reason: String) extends Confirmation
```

Caller sends a command, but `PersistentEntity` rejects it with: `replyTo ! Rejected("Invalid command")`.

Caller gets a `Future[Confirmation]` that actually contains a `Rejected`, but the future is still completed with a `Success`. 

The idea is to have the means to let the user inform Akka that the `Future` must fail. This can be done with a trait like

```scala
// defined in Akka
trait Error { def reason: String }

// defined by user
sealed trait Confirmation extends Response
case object Accepted extends Confirmation
case class Rejected(reason: String) extends Confirmation with Error
```

When completing the `Future` for a `ask`, Akka can inspect the value and detect that it's extending the marker trait `Error`. In that case, it completes the `Future` with a `Failure`.

**Comment by patriknw** [comment]
An alternative would be to let the receiver of the reply decide how to map it.

I think I prefer that because the error (rejection) messages can contain application specific info and mapping that automatically to an exception would loose the type of the error message.

**Comment by TimMoore** [comment]
Failing the `Future` on exception is how Lagom behaves now, and I think it's quite common to make use of this fact for domain validation.

If we want Lagom users to be able to migrate to Akka Persistence Typed (including a rolling update of a cluster from Lagom Persistent Entities to Akka Persistence Typed) we'll need to account for this in the migration guide. It sounds like some existing Lagom protocols can't be represented, so they'll need to first migrate their Lagom entities to a compatible protocol.

This is something else that will be pretty alien to many Java developers. It is a nice convenience that in Lagom, you can throw any exception from an entity command handler and (assuming it can be serialized) it will be received by the caller as a failed `Future`/`CompletionStage`.

**Comment by viktorklang** [comment]
Here's what I think (but I am open to be convinced of otherwise):
The problems that are a part of the business domain should be a part of the protocol algebra of the service. Ex: "InsufficientFunds"
The problems that fall outside of that should be surfaced as failed Futures (Timeouts, communication problems etc)

**Comment by TimMoore** [comment]
@viktorklang I don't think such a straightforward distinction is possible in the Akka model. For example, if the persistent entity fails to write an event because the database is down, is that part of the business domain? (Presumably: no) Would you expect it to immediately propagate back to the `Future` returned by `ask` as a failure with the original `SQLException` (or whatever)? IIUC this is not possible and it will instead result in the `Future` failing with `AskTimeoutException` after the timeout elapses. The only possibilities are a success with a value returned by the entity, or a timeout.

**Comment by viktorklang** [comment]
@TimMoore Notifying the caller of unexpected failures is always on a best-effort basis (as the return channel could be down). I think this is more along the distinction between "what is failure supervison and what is domain errors". Does that make more sense?

**Comment by TimMoore** [comment]
It all makes sense to me, I'm just pointing out that this is a change in allowed behavior vs. Lagom and Akka Untyped, which make a _better_ best effort to return failures. :smile:

**Comment by renatocaval** [comment]
@viktorklan, the user still needs to encode the error as part of the protocol. 

What we are saying is that we give the user the choice to say: when I return that domain error, `InsufficientFunds`, I want the `Future` to fail with that message.

Obviously, the same can be done with a `map` as @patriknw suggested. And that will still be possible and will be the default case. 

The only problem with that is that it puts the burden on the caller and this needs to be done over and over again creating unnecessary code and noise.

**Comment by patriknw** [comment]
> The only problem with that is that it puts the burden on the caller and this needs to be done over and over again creating unnecessary code and noise.

On the other hand we introduce a pattern to fail the Future that only works for simple response types, but another approach must be used for other more rich response types (which are also domain errors). More things to learn than the more clear distinction between system failures and domain errors.

Adding the choice to fail the future for certain reply messages also introduces the possibility that different actors implement this differently, possibly even for different commands. It would be encoded in the message types, which is good, but you have to look closer than if you know that all replies are always success values of the Futures.

**Comment by renatocaval** [comment]
@patriknw, still, I think that this is useful and make it much more convenient to use. 

> On the other hand we introduce a pattern to fail the Future that only works for simple response types, but another approach must be used for other more rich response types (which are also domain errors). 

This can easily be solved by extending a marker exception instead of a trait. As such we can preserve the type with all the information in it but still, fail the future.

> More things to learn than the more clear distinction between system failures and domain errors.

I don't think this is such an important thing. At the end of the day, when the command fails, we don't want to continue the sequence of computations. We want a failed Future to short-circuit a flatMap.

**Comment by viktorklang** [comment]
Is there a reasson why we wouldn't want to keep this symmetric with `ask`?

**Comment by renatocaval** [comment]
> Is there a reasson why we wouldn't want to keep this symmetric with `ask`?

I don't get your question.

**Comment by viktorklang** [comment]
@renatocaval Let me rephrase. Should there be any difference in the logic needed to be written to fail an (ordinary) `ask`-request?

**Comment by viktorklang** [comment]
@renatocaval Are you intending this as a replacement for akka.actor.Status?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala [hunk 7]
```diff
@@ -169,11 +168,31 @@ abstract class PersistentBehavior[Command, Event, State >: Null] private (val pe
...
+  def this(persistenceId: PersistenceId) = {
+    this(persistenceId, Optional.empty[BackoffSupervisorStrategy])
+  }
+
+  def this(persistenceId: PersistenceId, backoffSupervisorStrategy: BackoffSupervisorStrategy) = {
+    this(persistenceId, Optional.ofNullable(backoffSupervisorStrategy))
+  }
+
+  // FIXME override commandHandler and commandHandlerBuilder to require the ReplyEffect return type,
+  // which is unfortunately intrusive to the CommandHandlerBuilder
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala]: @renatocaval I continued implementing this more for real, in javadsl. It will require another type parameter to the CommandHandlerBuilder since it must use `Effect` or `ReplyEffect` as return type of the `matchCommand`. Makes it more complicated, but should be possible, just sharing progress.

**Comment by renatocaval** [comment]
> Are you intending this as a replacement for akka.actor.Status?

@viktorklang, good question. I actually I don't know how `akka.actor.Status` fits in Akka Typed. You can't send it to an `ActorRef[T]`, as far as I can say. 

@patriknw, can we send an untyped `Status` to an `ActorRef[T]` as a signal? Because in that case, we could have an `Effect.reject(msg)` method that signals it to fail the `Future`. That would be a better solution, IMO.

**Comment by patriknw** [comment]
Status can’t be used unless T is Status.

reject (and sending untyped Status.Failed that is not a T) can’t be used because it doesn’t have to be an ask involved. The command can be sent from another actor with tell.

The Error trait can technically work as a way to fail ask Futures. For me it’s more a question of it’s a pattern we want to encourage.

**Comment by viktorklang** [comment]
@patriknw @renatocaval I think the akka.actor.Status "replacement" is an important discussion to settle as well. So we don't introduce something specific where something generic is needed.

**Comment by patriknw** [comment]
@viktorklang I added ticket https://github.com/akka/akka/issues/25781

**Comment by patriknw** [comment]
@chbatey I added docs, and would like to merge this although the javadsl isn't not fully implemented yet. I don't want to conflict with the upcoming changes of the builders.

File: akka-docs/src/main/paradox/typed/persistence.md [hunk 8]
```diff
@@ -248,6 +248,35 @@ Java
...
+## Replies
+
+The @ref:[Request-Response interaction pattern](interaction-patterns.md#request-response) is very common for
+persistent actors, because you typically want to know if the command was rejected due to validation errors and
+when accepted you want a confirmation when the events have been successfully stored.
+
+Therefore you typically include a @scala[`ActorRef[ReplyMessageType]`]@java[`ActorRef<ReplyMessageType>`] in the
+commands. After validation errors or `thenRun` side effect the reply message can be sent to the `ActorRef`.
+
+TODO example of thenRun reply
```
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/persistence.md]: I want to use the examples from #25779 and therefore leaving this as a TODO, since that PR is on top of this PR.

File: akka-docs/src/main/paradox/typed/persistence.md [hunk 9]
```diff
@@ -248,6 +248,35 @@ Java
...
 The `SideEffect`s are executed sequentially, it is not possible to execute `SideEffect`s in parallel.
 
+## Replies
+
+The @ref:[Request-Response interaction pattern](interaction-patterns.md#request-response) is very common for
+persistent actors, because you typically want to know if the command was rejected due to validation errors and
+when accepted you want a confirmation when the events have been successfully stored.
+
+Therefore you typically include a @scala[`ActorRef[ReplyMessageType]`]@java[`ActorRef<ReplyMessageType>`] in the
+commands. After validation errors or `thenRun` side effect the reply message can be sent to the `ActorRef`.
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/persistence.md]: is this more or after it has been persisted using a thenRun side effect?

File: akka-docs/src/main/paradox/typed/persistence.md [hunk 10]
```diff
@@ -248,6 +248,35 @@ Java
...
+
+TODO example of thenRun reply
+
+Since this is such a common pattern there is reply effect for this purpose. It has the nice property that
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/persistence.md]: `is a reply effect`?

File: akka-docs/src/main/paradox/typed/persistence.md [hunk 11]
```diff
@@ -248,6 +248,35 @@ Java
...
+it can be used to enforce that replies are not forgotten when implementing the `PersistentBehavior`.
+If it's defined with @scala[`PersistentBehavior.withEnforcedReplies`]@java[`PersistentBehaviorWithEnforcedReplies`]
+there will be compilation errors if the returned effect isn't a `ReplyEffect`, which can be
+created with @scala[`Effect.reply`]@java[`Effects().reply`], @scala[`Effect.noReply`]@java[`Effects().noReply`],
+@scala[`Effect.thenReply`]@java[`Effects().thenReply`], or @scala[`Effect.thenNoReply`]@java[`Effects().thenNoReply`].
+
+These effects will send the reply message also when @scala[`PersistentBehavior.withEnforcedReplies`]@java[`PersistentBehaviorWithEnforcedReplies`]
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/persistence.md]: `even when` maybe reads better

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EffectImpl.scala [hunk 12]
```diff
@@ -4,37 +4,43 @@
 
 package akka.persistence.typed.internal
 
-import akka.persistence.typed.{ SideEffect, javadsl, scaladsl }
+import scala.collection.immutable
```
Comment (chbatey) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EffectImpl.scala]: nit: We use immutable.Seq and im.Seq in this file

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/Effect.scala [hunk 13]
```diff
@@ -40,6 +41,31 @@ object EffectFactory extends EffectFactories[Nothing, Nothing, Nothing]
...
+   *
+   * This has the same semantics as `cmd.replyTo.tell`.
+   *
+   * It is provided as a convenience (reducing boilerplate) and a way to enforce that replies are not forgotten
+   * when the `PersistentBehavior` is created with [[PersistentBehaviorWithEnforcedReplies]]. When
+   * `withEnforcedReplies` is used there will be compilation errors if the returned effect isn't a [[ReplyEffect]].
+   * The reply message will be sent also if `withEnforcedReplies` isn't used, but then the compiler will not help
+   * finding mistakes.
+   */
+  def reply[ReplyMessage](cmd: ExpectingReply[ReplyMessage], replyWithMessage: ReplyMessage): ReplyEffect[Event, State] =
```
Comment (chbatey) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/Effect.scala]: Should we add a TODO/FIXME that this isn't finished/tested yet

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala [hunk 14]
```diff
@@ -169,11 +168,33 @@ abstract class PersistentBehavior[Command, Event, State >: Null] private (val pe
...
+    if (supervisorStrategy.isPresent)
       behavior.onPersistFailure(supervisorStrategy.get)
     else
       behavior
   }
 
 }
 
+/**
+ * FIXME This is not completed for javadsl yet. The compiler is not enforcing the replies yet.
```
Comment (chbatey) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior.scala]: Ah here it is :)

**Review by chbatey — APPROVED** [review]
Maybe add a note to the docs that the Java API is still a work in progress in case we end up doing a release?

**Comment by patriknw** [comment]
thanks for comments, updated, and rebased
